src/common/getType.ts

```diff
@@ -400,57 +404,82 @@ function getOpCodeType(
     case "text_concat": {
       expectType(textType(), textType());
       const [t1, t2] = types as [TextType, TextType];
-      return textType(t1.capacity + t2.capacity);
+      return textType(
+        getArithmeticType("add", t1.codepointLength, t2.codepointLength),
+        t1.isAscii && t2.isAscii
+      );
     }
     case "repeat": {
       expectType(textType(), integerType(0));
       const [t, i] = types as [TextType, IntegerType];
-      return textType(mul(BigInt(t.capacity), i.high));
+      return textType(
+        getArithmeticType("mul", t.codepointLength, i),
+        t.isAscii
+      );
     }
     case "text_contains":
       expectType(textType(), textType());
       return booleanType;
-    case "text_find":
+    case "text_codepoint_find":
+    case "text_byte_find":
       expectType(textType(), textType());
-      return integerType(-1, (types[0] as TextType).capacity - 1);
+      return integerType(
+        -1,
+        mul(
+          sub((types[0] as TextType).codepointLength.high, 1n),
```

Comment:
```suggestion
          sub((types[0] as TextType).codepointLength.high, (types[1] as TextType).codepointLength.low),
```

Slightly narrower, probably doesn't matter.